# Case Study 2: T20I Match Analysis (2024)

## Context
This case study analyzes international T20 cricket match results recorded in 2024. The goal is to answer practical analytics questions such as head-to-head performance, team rankings, win rates, venue dominance, and margin-based trends using SQL.

## Schema
**Schema:** `case_study_2`  
**Table:** `T20I`

**Columns:**
- `team1`: First team in the fixture
- `team2`: Second team in the fixture
- `winner`: Match winner (`tied` / `no result` possible)
- `margin`: Winning margin as text (e.g., `35 runs`, `6 wickets`)
- `matchdate`: Match date
- `ground`: Venue/ground name

## Questions
The notebook answers 10 core business-style SQL questions, moving from simple filters to ranking, aggregation, and window-function based analysis.

## Setup: connect to Postgres

In [1]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

load_dotenv(dotenv_path='../.env')

DB_USER = os.getenv('DB_USER', 'postgres')
DB_PASSWORD = os.getenv('DB_PASSWORD', '')
DB_HOST = os.getenv('DB_HOST', 'localhost')
DB_PORT = os.getenv('DB_PORT', '5432')
DB_NAME = os.getenv('DB_NAME', 'sql_case_studies')

CONN_STR = f'postgresql+psycopg2://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}'
engine = create_engine(CONN_STR)
print('Connecting to:', f'postgresql://{DB_USER}:***@{DB_HOST}:{DB_PORT}/{DB_NAME}')

Connecting to: postgresql://postgres:***@localhost:5432/sql_case_studies


In [2]:
%load_ext sql
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False
%sql engine

The 'toml' package isn't installed. To load settings from pyproject.toml or ~/.jupysql/config, install with: pip install toml

In [3]:
%%sql
SELECT version();

,version
0,"PostgreSQL 17.6 on x86_64-windows, compiled by..."


In [10]:
%%sql
SET search_path = case_study_2;
SELECT COUNT(*) AS total_rows
FROM T20I;

,total_rows
0,122


In [14]:
%%sql
SELECT *
FROM T20I
LIMIT 5;

,team1,team2,winner,margin,matchdate,ground
0,West Indies,England,West Indies,5 wickets,2024-11-16,Gros Islet
1,Australia,Pakistan,Australia,13 runs,2024-11-16,Sydney
2,South Africa,India,India,135 runs,2024-11-15,Johannesburg
3,West Indies,England,England,3 wickets,2024-11-14,Gros Islet
4,Australia,Pakistan,Australia,29 runs,2024-11-14,Brisbane


----
## Question 1: Identify matches played between two specific teams (e.g., India and South Africa) in 2024 and their results.

### Thought process
First, confirm the available date range to verify whether a year filter is needed. Then filter rows where the two teams are exactly India and South Africa (in either column) and select only matchup + result fields for a clean head-to-head view.

In [17]:
%%sql
SELECT MIN(matchdate), MAX(matchdate)
FROM T20I;

,min,max
0,2024-01-11,2024-11-16


So our date range goes from 11/01/2024 upto 16/11/2024, all records are in the year of 2024. No year filter needed.

In [19]:
%%sql
SELECT
    team1
    , team2
    , winner
FROM T20I
WHERE team1 IN ('India', 'South Africa') AND team2 IN ('India', 'South Africa');

,team1,team2,winner
0,South Africa,India,India
1,South Africa,India,India
2,South Africa,India,South Africa
3,South Africa,India,India
4,India,South Africa,India


### Insight
India dominated this 2024 head-to-head, winning 4 of the 5 listed matches against South Africa. The result distribution is clearly one-sided in India’s favor for this pairing.

---
## Question 2: Find the team with the highest number of wins in 2024 and the total matches it won.

### Thought process
Since all rows are from 2024, group by `winner`, count wins, sort descending, and keep the top row. This directly returns the most successful team by total wins.

In [23]:
%%sql
SELECT
    winner
    , COUNT(*) AS matches_won
FROM T20I
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1;

,winner,matches_won
0,India,22


### Insight
India appears as the top winner with **22 matches won**, making it the most successful team in this dataset for 2024.

----
## Question 3: Rank the teams based on the total number of wins in 2024.

### Thought process
Build a pre-aggregated table of wins per team (excluding `tied` and `no result`), then apply a window `RANK()` by descending wins to produce leaderboard-style output and preserve ties.

In [28]:
%%sql
WITH teams_total_wins AS
    (SELECT
        winner
        , COUNT(*) AS matches_won
    FROM T20I
    WHERE winner NOT IN ('tied', 'no result')
    GROUP BY 1)

SELECT
    *
    , RANK() OVER(ORDER BY matches_won DESC) AS rank
FROM teams_total_wins
;

,winner,matches_won,rank
0,India,22,1
1,Australia,15,2
2,West Indies,12,3
3,Sri Lanka,9,4
4,England,9,4
5,South Africa,9,4
6,Bangladesh,8,7
7,New Zealand,7,8
8,Pakistan,7,8
9,Afghanistan,6,10


### Insight
India leads the win rankings (22). Australia is second (15), and there is a tie at rank 4 among Sri Lanka, England, and South Africa (9 each), which validates the use of `RANK()` for tie-aware standings.

---
## Question 4: Which team had the highest average winning margin (in runs), and what was the average margin?

### Thought process
Inspect sample rows first to confirm `margin` format, then isolate `runs` results, extract the numeric component with `split_part`, and compare team-level averages against the overall average for context.

In [29]:
%%sql
SELECT * FROM T20I LIMIT 5;

,team1,team2,winner,margin,matchdate,ground
0,West Indies,England,West Indies,5 wickets,2024-11-16,Gros Islet
1,Australia,Pakistan,Australia,13 runs,2024-11-16,Sydney
2,South Africa,India,India,135 runs,2024-11-15,Johannesburg
3,West Indies,England,England,3 wickets,2024-11-14,Gros Islet
4,Australia,Pakistan,Australia,29 runs,2024-11-14,Brisbane


In [ ]:
%%sql
WITH overall_avg_margin AS
    (SELECT
        ROUND(AVG(split_part(margin, ' ', 1)::INTEGER), 2) AS overall_avg_margin
    FROM T20I
    WHERE margin LIKE '%runs')


SELECT
    winner
    , ROUND(AVG(split_part(margin, ' ', 1)::INTEGER), 2) AS winner_avg_margin
    , overall_avg_margin.overall_avg_margin
FROM T20I
CROSS JOIN overall_avg_margin
WHERE margin LIKE '%runs'
GROUP BY 1, overall_avg_margin.overall_avg_margin
ORDER BY winner_avg_margin DESC
LIMIT 1;

,winner,winner_avg_margin,overall_avg_margin
0,India,55.73,34.19


### Insight
India has the highest average winning margin by runs at **55.73**, well above the overall runs-margin average of **34.19**. This suggests India’s wins-by-runs were not only frequent but also comparatively dominant.

## Question 4.1: Which team had the highest average winning margin (in wickets), and what was the average margin?

### Thought process
Repeat the same margin parsing pattern but restrict to `wickets` outcomes. This keeps the metric consistent while isolating chase-style victories.

In [48]:
%%sql
WITH overall_avg_margin AS
    (SELECT
        ROUND(AVG(split_part(margin, ' ', 1)::INTEGER), 2) AS overall_avg_margin
    FROM T20I
    WHERE margin LIKE '%wickets')


SELECT
    winner
    , ROUND(AVG(split_part(margin, ' ', 1)::INTEGER), 2) AS winner_avg_margin
    , overall_avg_margin.overall_avg_margin
FROM T20I
CROSS JOIN overall_avg_margin
WHERE margin LIKE '%wickets'
GROUP BY 1, overall_avg_margin.overall_avg_margin
ORDER BY winner_avg_margin DESC
LIMIT 1;

,winner,winner_avg_margin,overall_avg_margin
0,India,7.29,6.48


### Insight
India also leads in average wickets-margin wins at **7.29 wickets**, slightly above the overall wickets average of **6.48**. This indicates strong finishing efficiency in successful run chases.

---
## Question 5: List all matches where the winning margin was greater than the average margin across all matches. (in runs)

### Thought process
Compute the global average runs margin first, then return only matches whose numeric runs margin exceeds that baseline. This identifies outlier-level dominant wins.

In [74]:
%%sql
SELECT
    *
    , split_part(margin, ' ', 1)::INTEGER AS winning_margin
FROM T20I
WHERE margin LIKE '%runs' AND split_part(margin, ' ', 1)::INTEGER  >  (
    (SELECT
        ROUND(AVG(split_part(margin, ' ', 1)::INTEGER), 2) AS average_margin
    FROM T20I
    WHERE margin LIKE '%runs')
);

,team1,team2,winner,margin,matchdate,ground,winning_margin
0,South Africa,India,India,135 runs,2024-11-15,Johannesburg,135
1,South Africa,India,India,61 runs,2024-11-08,Durban,61
2,Sri Lanka,West Indies,Sri Lanka,73 runs,2024-10-15,Dambulla,73
3,India,Bangladesh,India,133 runs,2024-10-12,Hyderabad,133
4,India,Bangladesh,India,86 runs,2024-10-09,Delhi,86
5,Scotland,Australia,Australia,70 runs,2024-09-06,Edinburgh,70
6,Sri Lanka,India,India,43 runs,2024-07-27,Pallekele,43
7,Zimbabwe,India,India,42 runs,2024-07-14,Harare,42
8,Zimbabwe,India,India,100 runs,2024-07-07,Harare,100
9,England,India,India,68 runs,2024-06-27,Providence,68


### Insight
The output highlights several high-separation wins (e.g., 135, 133, and 104 runs), with many involving India. These are matches where the victory margin is meaningfully above the average runs-margin benchmark.

---
## Question 6: Find the team with the most wins when chasing a target (wins by wickets)

### Thought process
Filter to `margin LIKE '%wickets'` to capture chasing wins, aggregate by `winner`, then rank descending to identify the team(s) with the maximum count.

In [88]:
%%sql
WITH ranking_table AS 
    (SELECT
        winner
        , COUNT(*) AS matches_won_by_wickets
        , RANK() OVER(ORDER BY COUNT(*) DESC)
    FROM T20I
    WHERE margin LIKE '%wickets'
    GROUP BY 1
    ORDER BY 2 DESC) 

SELECT 
    winner
    , matches_won_by_wickets
FROM ranking_table
WHERE rank = 1;


,winner,matches_won_by_wickets
0,England,7
1,India,7


### Insight
The result is a tie: **England** and **India** both recorded **7 wins by wickets**, so there is no single leader for chasing wins in this dataset.

---
## Question 7: Head-to-head record between two selected teams (e.g., England vs Australia).

### Thought process
Parameterize the two teams in a CTE so the same query can be reused. Then filter both team-order combinations and group by `winner` to get the head-to-head win split.

In [94]:
%%sql
-- #Change the params as needed
WITH params AS (
    SELECT
        'England'::varchar AS team_a
        , 'Australia'::varchar AS team_b
)

SELECT
    winner
    , COUNT(*) AS matches_won
FROM T20I CROSS JOIN params
WHERE (team1=team_a AND team2=team_b) OR (team1=team_b AND team2=team_a)
GROUP BY 1

,winner,matches_won
0,Australia,2
1,England,1


### Insight
For the selected pair (England vs Australia), Australia leads the head-to-head in this dataset with **2 wins** versus **1** for England.

---
## Question 8: Identify the month in 2024 with the highest number of T20I matches played.

### Thought process
Extract month from `matchdate`, count matches per month, and sort descending to find the busiest month in one step.

In [104]:
%%sql
SELECT
    date_part('month', matchdate)::INTEGER AS month
    , COUNT(*) AS matches
FROM T20I
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1;

,month,matches
0,6,36


### Insight
Month **6 (June)** has the highest activity with **36 matches**, likely reflecting a major tournament period in the 2024 calendar.

---
## Question 9: For each team, find how many matches they played in 2024 and their win percentage.

### Thought process
Count matches played by each team by unioning `team1` and `team2`, aggregate totals, then left join with wins-per-team and compute win percentage with null-safe division.

In [141]:
%%sql
WITH union_table AS 
    (SELECT
        team1 AS team
        , COUNT(*) AS number_of_matches
    FROM T20I
    GROUP BY 1
    UNION ALL
    SELECT
        team2 AS team
        , COUNT(*) AS number_of_matches
    FROM T20I
    GROUP BY 1)

, team_matches_count AS
    (SELECT  
        team
        , SUM(number_of_matches) AS matches_count
    FROM union_table
    GROUP BY 1)

, win_table AS
    (SELECT
        winner
        , COUNT(*) AS matches_won
    FROM T20I
    WHERE winner NOT IN ('tied', 'no result')
    GROUP BY 1)

SELECT
    team
    , matches_count
    , ROUND(100 * COALESCE(matches_won::DECIMAL / matches_count, 0), 1) AS win_percentage
FROM team_matches_count t LEFT JOIN win_table w
ON t.team = w.winner

,team,matches_count,win_percentage
0,Bangladesh,20,40.0
1,Zimbabwe,13,23.1
2,New Zealand,17,41.2
3,U.S.A.,10,40.0
4,Scotland,5,0.0
5,Pakistan,21,33.3
6,Australia,19,78.9
7,Ireland,11,27.3
8,Sri Lanka,19,47.4
9,Canada,6,50.0


### Insight
India stands out with **26 matches** and an **84.6%** win rate (highest among heavily active teams). Australia is also strong at **78.9%**, while some teams (e.g., Scotland, Namibia, Nepal) show **0%** in this sample period.

---
## Question 10: Identify the most successful team at each ground (team with most wins per ground).

### Thought process
Validate table structure, then count wins by `ground` and `winner`. Use `RANK() OVER (PARTITION BY ground ORDER BY wins DESC)` so ties at a venue are preserved rather than arbitrarily dropped.

In [142]:
%%sql
SELECT *
FROM T20I
LIMIT 5;

,team1,team2,winner,margin,matchdate,ground
0,West Indies,England,West Indies,5 wickets,2024-11-16,Gros Islet
1,Australia,Pakistan,Australia,13 runs,2024-11-16,Sydney
2,South Africa,India,India,135 runs,2024-11-15,Johannesburg
3,West Indies,England,England,3 wickets,2024-11-14,Gros Islet
4,Australia,Pakistan,Australia,29 runs,2024-11-14,Brisbane


In [154]:
%%sql
with rankings AS 
    (SELECT *, RANK() OVER(PARTITION BY ground ORDER BY wins DESC) AS rank
    FROM 
        (SELECT 
            ground
            , winner
            , COUNT(*) AS wins
        FROM T20I 
        WHERE winner NOT IN ('tied','no result')
        GROUP BY 1, 2))

SELECT
    ground
    , winner
    , wins
FROM rankings
WHERE rank = 1;

,ground,winner,wins
0,Abu Dhabi,Ireland,1
1,Abu Dhabi,South Africa,1
2,Adelaide,Australia,1
3,Auckland,Australia,2
4,Birmingham,England,1
5,Bridgetown,England,3
6,Brisbane,Australia,1
7,Cardiff,England,1
8,Centurion,India,1
9,Chattogram,Bangladesh,3


### Insight
The output correctly returns one or more top teams per ground because ties are retained. For example, some venues show a single clear leader (e.g., Dambulla: Sri Lanka, Harare: India), while others have co-leaders with equal win counts.